accounting/invoice/tests/test_views.py

```diff
@@ -20,3 +20,71 @@
 """
 Tests for Invoice views.
 """
+
+import datetime
+
+from django.utils import timezone
+from rest_framework import status
+from rest_framework.reverse import reverse
+from rest_framework.test import APITestCase
+import freezegun
+
+from accounting.account.tests.factories import AccountFactory
+from accounting.authentication.tests.factories import UserFactory
+from accounting.invoice import choices
+from accounting.invoice.tests import factories
+
+NOW = datetime.datetime(2018, 1, 10, 20, 31, 3, 350993, tzinfo=timezone.utc)
+
+
+class InvoiceViewSetTestCase(APITestCase):
+    """ Test cases for `views.InvoiceViewSet`. """
+
+    @freezegun.freeze_time(NOW)
+    def setUp(self):
+        """ Set up test objects. """
+        self.admin = UserFactory(is_staff=True)
+        self.provider_account = AccountFactory(business_name='Developer', user__username='developer')
+        self.client_account = AccountFactory(business_name='OpenCraft GmbH', user__username='opencraft')
+        self.invoice = factories.InvoiceFactory(provider=self.provider_account, client=self.client_account, date=NOW)
+        super().setUp()
+
+    def test_approve(self):
+        """ Approving an invoice through the appropriate URL is considered manual approval. """
+        # First visit approves the invoice.
+        url = reverse('invoice:invoice-approve', [self.invoice.uuid])
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.data, {'approved': True})
+        self.invoice.refresh_from_db()
+        self.assertTrue(self.invoice.is_approved)
+        self.assertEqual(self.invoice.approved, choices.InvoiceApproval.manually)
+
+        # Second visit when already approved doesn't change anything.
+        url = reverse('invoice:invoice-approve', [self.invoice.uuid])
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
```

Comment:
@UmanShahzad Hm... I'm not sure if 404 is the right status code to use here. After all, the resource still exists. Perhaps a 403 (Forbidden) that includes some info about why the server is refusing action would be more appropriate?

Reply:
Actually, it makes sense given how the view handles this: when trying to `approve` or `pay` an invoice, you can only do that for invoices that aren't already approved or paid, respectively. So the view only filters invoices with the appropriate condition, i.e. not approved or not paid. So if you get a 404 it means it could not find an invoice for approval or an invoice that's still payable. Of course if you remove the verb, it will find it (if you have admin access).